lib/corefoundation/preferences.rb

```diff
@@ -23,49 +20,133 @@ module CF
-        CF::Base.typecast(plist_ref).to_ruby unless plist_ref.null?
-      end
+    # Set the value for preference domain using `CFPreferencesSetValue`.
+    #
+    # @param [String] key Preference key
+    # @param [Integer, Float, String, TrueClass, FalseClass, Hash, Array] value Preference value
+    # @param [String] application_id Preference domain
+    # @param [String, Symbol] username Domain user (current, any or a specific user)
+    # @param [String, Symbol] hostname Hostname (current, all hosts or a specific host)
+    #
+    # @return [TrueClass, FalseClass] Returns true if preference was successfully written to storage, otherwise false.
+    #
+    def self.set(key, value, application_id, username = nil, hostname = nil)
+      username ||= CURRENT_USER
+      hostname ||= ALL_HOSTS
+      CF.CFPreferencesSetValue(key.to_cf, arg_to_cf(value), application_id.to_cf, arg_to_cf(username), arg_to_cf(hostname))
+      CF.CFPreferencesAppSynchronize(application_id.to_cf)
+    end
 
-      def get!(key, application_id, username = nil, hostname = nil)
-        get(key, application_id, username, hostname) || (raise CF::Error, "Preference key `#{key}` not found")
-      end
+    # Calls the {#self.set} method and raise a `PreferenceSyncFailed` error if `false` is returned.
+    #
+    # @param [String] key Preference key to write
+    # @param [String] application_id Preference domain for the key
+    # @param [String, Symbol] username Domain user (current, any or a specific user)
+    # @param [String, Symbol] hostname Hostname (current, all hosts or a specific host)
+    #
+    # @raise [PreferenceSyncFailed] If {#self.set} call returned false.
+    #
+    # @return [VALUE] Returns true if preference value is successfully written.
+    #
+    def self.set!(key, value, application_id, username = nil, hostname = nil)
+      hostname = arg_to_cf(hostname || ALL_HOSTS)
+      hostname = CF::Base.typecast(hostname).to_ruby
+      raise(CF::Exceptions::PreferenceSyncFailed.new(key, application_id, hostname)) unless set(key, value, application_id, username, hostname)
+    end
 
-      def set(key, value, application_id, username = CURRENT_USER, hostname = CURRENT_HOST)
-        CF.CFPreferencesSetValue(key.to_cf, value.to_cf, application_id.to_cf, arg_to_cf(username), arg_to_cf(hostname))
-        CF.CFPreferencesAppSynchronize(application_id.to_cf)
-      end
+    # Checks whether a key exists in the preference domain. It'll also return false for an invalid domain.
+    #
+    # @param [String] key Preference key to read
+    # @param [String] application_id Preference domain for the key
+    # @param [String, Symbol] username Domain user (current, any or a specific user)
+    # @param [String, Symbol] hostname Hostname (current, all hosts or a specific host)
+    #
+    # @return [TrueClass, FalseClass] Return true or false to indicate if it is a valid key.
+    #
+    def self.valid_key?(key, application_id, username = nil, hostname = nil)
```

Comment:
If this method has no usage in here or `macos_userdefaults` resource, let's chuck this.

Reply:
This has been added only because it is needed. Not point in putting something we're not going to use. 😄

Reply:
Seems, like this is being used in `macos_userdefaults` resource. I suggest we remove this altogether for performance reasons and simplicity as this is listing keys and finding in ruby again, when the CF function `CFPreferencesCopyValue` returns `null` anyway if the value is not located

Reply:
Yeah I guess we could do that. Let me try to remember why it was necessary to check if it is a valid key.

Reply:
Can't remember why I kept that key check in `macos_userdefaults` and didn't make sense to keep it so removed that and these two methods too since they are not necessary anymore. Resolving this conversation. Thanks!